clippy_lints/src/methods/slow_manual_contains.rs

```diff
@@ -0,0 +1,153 @@
+use clippy_utils::msrvs::{Msrv, SLICE_CONTAINS_SIMD};
+use clippy_utils::source::snippet_with_applicability;
+use clippy_utils::{eq_expr_value, peel_hir_pat_refs};
+use rustc_ast::UnOp;
+use rustc_errors::Applicability;
+use rustc_hir::def::Res;
+use rustc_hir::{BinOpKind, Body, Expr, ExprKind, HirId, QPath};
+use rustc_lint::LateContext;
+use rustc_middle::ty::{self};
+use rustc_span::Span;
+use rustc_span::source_map::Spanned;
+
+use super::SLOW_MANUAL_CONTAINS;
+
+pub(super) fn check(cx: &LateContext<'_>, expr: &Expr<'_>, recv: &Expr<'_>, closure_arg: &Expr<'_>, msrv: &Msrv) {
+    let mut app = Applicability::MachineApplicable;
+
+    if !expr.span.from_expansion()
+        // check if `iter().any()` can be replaced with `contains()`
+        && let ExprKind::Closure(closure) = closure_arg.kind
+        && let Body{params: [param],value} = cx.tcx.hir().body(closure.body)
+        && let ExprKind::Binary(op, lhs, rhs) = value.kind
+        && let (peeled_ref_pat, _)=peel_hir_pat_refs(param.pat)
+        && let (true, snip)=can_replace_with_contains(cx, op, lhs, rhs, peeled_ref_pat.hir_id, &mut app)
+    {
+        let ref_type = cx.typeck_results().expr_ty_adjusted(recv);
+
+        match ref_type.kind() {
+            ty::Ref(_, inner_type, _) if inner_type.is_slice() => {
+                // check if the receiver is a numeric slice
+                if let ty::Slice(slice_type) = inner_type.kind()
+                    && ((slice_type.to_string() == "u8" || slice_type.to_string() == "i8")
+                        || (slice_type.is_numeric() && msrv.meets(SLICE_CONTAINS_SIMD)))
+                {
+                    span_lint_and_sugg(
+                        cx,
+                        SLOW_MANUAL_CONTAINS,
+                        expr.span,
+                        "using `contains()` instead of `iter().any()` is more efficient",
+                        "try",
+                        format!(
+                            "{}.contains({})",
+                            snippet_with_applicability(cx, recv.span, "..", &mut app),
+                            snip
+                        ),
+                        app,
+                    );
+                }
+            },
+            _ => {},
+        }
+    }
+}
+
+// Returns `true` and the snippet of the argument to `contains()` if the `iter().any()` call can be
+// replaced with `contains()`
+fn can_replace_with_contains<'tcx>(
+    cx: &LateContext<'tcx>,
+    bin_op: Spanned<BinOpKind>,
+    left_expr: &'tcx Expr<'tcx>,
+    right_expr: &'tcx Expr<'tcx>,
+    closure_arg_id: HirId,
+    applicability: &mut Applicability,
+) -> (bool, String) {
+    let mut get_snippet = |span: Span, needs_borrow: bool| {
+        let snippet = snippet_with_applicability(cx, span, "..", applicability).to_string();
+
+        if needs_borrow { format!("&{snippet}") } else { snippet }
+    };
+
+    match (bin_op.node, &left_expr.kind, &right_expr.kind) {
```

Comment:
This looks very repetitive. Can't you use an helper function which handles one argument at a time and returns an enumeration `Option<EligibleArg>` and combine the results? If `can_replace_with_contains()` returns an `Option<Span>` then you can even use `?` to return early.

```rust
enum EligibleArg {
    IsClosureArg,
    ContainsArg(String),
}
```